core/src/main/scala/kafka/tools/DumpLogSegments.scala

```diff
@@ -150,6 +173,76 @@ object DumpLogSegments {
+    val timeIndex = new TimeIndex(file, baseOffset = startOffset)
+
+    //Check that index passes sanityCheck, this is the check that determines if indexes will be rebuilt on startup or not.
+    if (indexSanityOnly) {
+      timeIndex.sanityCheck
+      println(s"$file passed sanity check.")
+      return
+    }
+
+    var prevTimestamp = Message.NoTimestamp
+    for(i <- 0 until timeIndex.entries) {
+      val entry = timeIndex.entry(i)
+      val position = index.lookup(entry.offset).position
+      var partialFileMessageSet: FileMessageSet = messageSet.read(position, Int.MaxValue)
+      val shallowIter = partialFileMessageSet.iterator
+      var maxTimestamp = Message.NoTimestamp
+      var checkedOffset = -1L
+      // We stop scan when one of the following is not met:
+      // 1. All the messages are checked
+      // 2. The max timestamp checked is greater or equals to the indexed timestamp
+      // 3. When the offset checked is greater than the next indexed offset.
+      var wrapperMessageOpt = if (shallowIter.hasNext) Some(shallowIter.next) else None
+      do {
+        wrapperMessageOpt match {
+          case Some(wrapperMessage) =>
+            val deepIter = getIterator(wrapperMessage, isDeepIteration = true)
+            for (messageAndOffset <- deepIter)
+              maxTimestamp = math.max(maxTimestamp, messageAndOffset.message.timestamp)
+            checkedOffset = wrapperMessage.offset
+            wrapperMessageOpt = if (shallowIter.hasNext) Some(shallowIter.next) else None
+          case _ =>
+        }
+      } while (wrapperMessageOpt.isDefined && maxTimestamp < entry.timestamp &&
+           (i == timeIndex.entries - 1 || timeIndex.entry(i + 1).offset > checkedOffset))
+
+      if (maxTimestamp != entry.timestamp) {
+        var misMatchesSeq = timeIndexDumpErrors.misMatchesForTimeIndexFilesMap.getOrElse(file.getAbsolutePath, List[(Long, Long)]())
+        misMatchesSeq ::=(entry.timestamp, maxTimestamp)
+        timeIndexDumpErrors.misMatchesForTimeIndexFilesMap.put(file.getAbsolutePath, misMatchesSeq)
+      }
+
+      if (prevTimestamp > entry.timestamp) {
+        var outOfOrderSeq = timeIndexDumpErrors.outOfOrderTimestamp.getOrElse(file.getAbsolutePath, List[(Long, Long)]())
+        outOfOrderSeq ::= (entry.timestamp, prevTimestamp)
+        timeIndexDumpErrors.outOfOrderTimestamp.put(file.getAbsolutePath, outOfOrderSeq)
+      }
+
+      // since it is a sparse file, in the event of a crash there may be many zero entries, stop if we see one
+      if(entry.offset == 0 && i > 0)
+        return
+      if (!verifyOnly)
+        println("timestamp: %s offset: %s".format(
+            entry.timestamp + { if (maxTimestamp == entry.timestamp) "" else s"(log timestamp: $maxTimestamp)" },
+            (timeIndex.baseOffset + entry.offset) +
+            { if (partialFileMessageSet.head.offset == entry.offset) "" else s"(log offset: ${partialFileMessageSet.head.offset})" }))
```

Comment:
I was thinking this easier for user to find out what when wrong. When I run this command, I found it is a little ugly that we have to switch between the dumped errors and the printed index entries.